clash-cores/src/Clash/Cores/SPI.hs

```diff
@@ -26,6 +32,53 @@ import Clash.Sized.Internal.BitVector
 
 import Clash.Cores.LatticeSemi.ICE40.IO
 import Clash.Cores.LatticeSemi.ECP5.IO
+import Clash.Class.HasDomain
+
+-- | Input signals of an SPI subordinate.
+data SpiSlaveIn (ds :: BiSignalDefault) (dom :: Domain) (misoLanes :: Nat) (mosiLanes :: Nat) =
+  SpiSlaveIn {
+    spiSlaveMosi :: "MOSI" ::: Signal dom (BitVector mosiLanes),
+      -- ^ Master-out, slave-in
+    spiSlaveMisoIn :: "MISO" ::: BiSignalIn ds dom misoLanes,
+      -- ^ Master-in, slave-out
+    spiSlaveSck :: "SCK" ::: Signal dom Bit,
+      -- ^ Serial Clock
+    spiSlaveSs :: "SS" ::: Signal dom Bit
+      -- ^ Slave select
+  }
+
+type instance TryDomain t (SpiSlaveIn ds dom misoLanes mosiLanes) = 'Found dom
+
+-- | Output signals of an SPI subordinate.
+data SpiSlaveOut (ds :: BiSignalDefault) (dom :: Domain) (misoLanes :: Nat) (mosiLanes :: Nat) =
+  SpiSlaveOut {
+    spiSlaveMisoOut :: "MISO" ::: BiSignalOut ds dom misoLanes
+      -- ^ Master-in, slave-out
+  }
+
+type instance TryDomain t (SpiSlaveOut ds dom misoLanes mosiLanes) = 'Found dom
+
+-- | Output signals of an SPI master.
+data SpiMasterOut (dom :: Domain) (misoLanes :: Nat) (mosiLanes :: Nat) =
+  SpiMasterOut {
+    spiMasterMosi :: "MOSI" ::: Signal dom (BitVector mosiLanes),
```

Comment:
So the data inputs and outputs of `spiMaster` are of type `Vec mosiW (BitVector n)` and `Vec misoW (BitVector n)` respectively, but `MOSI` is defined as `BitVector mosiLanes` and `MISO` as `BitVector misoLanes`.

The first thing that I think when I see this is: wait, a `Vec` is numbered with ascending inices ([`!!`](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Sized-Vector.html#v:-33--33-)) and a `BitVector` is numbered with descending indices ([`!`](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Class-BitPack-BitIndex.html#v:-33-))! So... how do those indices map?

And even if they were the same, I'd still wonder about the mapping to HDL indices. Does `din !! 0` map to `MOSI[0]` or `MOSI[n]` in my SDC file?

However, I think `BitVector _` for MOSI/MISO does have an advantage: it doesn't give the annoying
```
[...]: warning:
    Make sure HDL port names are correct:
Backtracked when constructing Clash.Sized.Vector.Vec
(Type appears recursive)
```

I'm not really sure about the type. We should clearly document the mapping of `Vec` indices to pins on the top entity, and ideally they'd match, so `din !! 0` does map to `MOSI[0]`. But whether `MOSI` should be `BitVector` or `Vec _ Bit` I'm not so sure about.